### ISSUE TITLE
chain/chord not propagating exceptions as expected

### ISSUE THREAD

**Issue by zaro** [celery/celery#3743]
## Checklist

- [x] I have included the output of ``celery -A proj report`` in the issue.
      (if you are not able to do this, then at least specify the Celery
       version affected).
- [ ] I have verified that the issue exists against the `master` branch of Celery.

## Steps to reproduce

Tasks:
```python
@app.task
def a1(arg):
    print('Task A1, starring with arg:' + str(arg))
    if arg == 3:
        raise Exception('A1 Error')
    time.sleep(5)
    print('A1 done')
    return {'a1': True}

@app.task
def a2(arg):
    print('Task A2, starring with arg:' + str(arg))
    time.sleep(5)
    print('A2 done')
    return {'a2': True}

@app.task
def b1(arg):
    print('Task B1, starring with arg:' + str(arg))
    time.sleep(5)
    print('B1 done')
    return {'b1': True}

```

If we start a chrod() with 3 a1 task fails as expected:
```python
>>> chord([ a1.s(1)  , a1.s(2) , a1.s(3)  ])(b1.s()).get()
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/Users/zaro/dev/proj-dev/proj/.direnv/python-3.5.2/lib/python3.5/site-packages/celery/result.py", line 189, in get
    on_message=on_message,
  File "/Users/zaro/dev/proj-dev/proj/.direnv/python-3.5.2/lib/python3.5/site-packages/celery/backends/async.py", line 191, in wait_for_pending
    return result.maybe_throw(callback=callback, propagate=propagate)
  File "/Users/zaro/dev/proj-dev/proj/.direnv/python-3.5.2/lib/python3.5/site-packages/celery/result.py", line 294, in maybe_throw
    self.throw(value, self._to_remote_traceback(tb))
  File "/Users/zaro/dev/proj-dev/proj/.direnv/python-3.5.2/lib/python3.5/site-packages/celery/result.py", line 287, in throw
    self.on_ready.throw(*args, **kwargs)
  File "/Users/zaro/dev/proj-dev/proj/.direnv/python-3.5.2/lib/python3.5/site-packages/vine/promises.py", line 213, in throw
    reraise(type(exc), exc, tb)
  File "/Users/zaro/dev/proj-dev/proj/.direnv/python-3.5.2/lib/python3.5/site-packages/vine/five.py", line 176, in reraise
    raise value
celery.backends.base.ChordError: Callback error: ChordError("Dependency e90cc09c-47bc-4fac-9a3d-08f1260c2d5c raised Exception('A1 Error',)",)
```

If the we use chains instead of a single task in the the chord like this:
```python
>>> chord([ ( a1.s(1) | a2.s()) , ( a1.s(2) | a2.s()), ( a1.s(3) | a2.s())  ])(b1.s()).get()

.... Hangs forever, and never throws ....
```

So first I though maybe chain() doesn't support error propagation, so I tried:

```python
>>> ( a1.s(3) | a2.s()).apply_async().get() 
.... Hangs forever, and never throws ....
```
And indeed it seems the behaviour is the same on a single chain.
But the I don't know why I tried it also with a variable:

```python
>>> r = ( a1.s(3) | a2.s()).apply_async()
>>> r.get()
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/Users/zaro/dev/proj-dev/proj/.direnv/python-3.5.2/lib/python3.5/site-packages/celery/result.py", line 172, in get
    self._maybe_reraise_parent_error()
  File "/Users/zaro/dev/proj-dev/proj/.direnv/python-3.5.2/lib/python3.5/site-packages/celery/result.py", line 195, in _maybe_reraise_parent_error
    node.maybe_throw()
  File "/Users/zaro/dev/proj-dev/proj/.direnv/python-3.5.2/lib/python3.5/site-packages/celery/result.py", line 294, in maybe_throw
    self.throw(value, self._to_remote_traceback(tb))
  File "/Users/zaro/dev/proj-dev/proj/.direnv/python-3.5.2/lib/python3.5/site-packages/celery/result.py", line 287, in throw
    self.on_ready.throw(*args, **kwargs)
  File "/Users/zaro/dev/proj-dev/proj/.direnv/python-3.5.2/lib/python3.5/site-packages/vine/promises.py", line 213, in throw
    reraise(type(exc), exc, tb)
  File "/Users/zaro/dev/proj-dev/proj/.direnv/python-3.5.2/lib/python3.5/site-packages/vine/five.py", line 176, in reraise
    raise value
celery.backends.base.Exception: A1 Error
```

And this chain works as expected.

So I tried the same thing with the chord():
```python
>>> r = chord([ ( a1.s(1) | a2.s()) , ( a1.s(2) | a2.s()), ( a1.s(3) | a2.s())  ])(b1.s())
>>> r.get()
.... Hangs forever, and never throws ....
```
but it didn't help.

## Expected behavior

While not being a python expert I would expect that:

```python
>>> ( a1.s(3) | a2.s()).apply_async().get() 
```
```python
>>> r = ( a1.s(3) | a2.s()).apply_async()
>>> r.get()
```
produce the same result

I would also expect that chord() will propagate exceptions regardless of whether it's header is list of tasks or list of chains().

## Actual behavior

chain() w/o intermediate variable doesn't propagate exceptions
chord() doesn't propagate exceptions occurring in the header chains.

## celery report

```

software -> celery:4.0.2 (latentcall) kombu:4.0.2 py:3.5.2
            billiard:3.5.0.2 redis:2.10.5
platform -> system:Darwin arch:64bit imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:redis results:redis:///

.......
CELERY_BROKER_URL: 'redis://localhost:6379//'
CELERY_RESULT_BACKEND: 'redis:///'
......
```

**Comment by mochawich** [comment]
I'm facing the same issue, @zaro have you figured out a way around it?

**Comment by zaro** [comment]
@mochawich unfortunately no . I am thinking to test with AMQP to see if it is also affected.

**Comment by msmfco** [comment]
I am facing the same issue. I couldn't find any way to handle this correctly using celery so I wrote a new base class that propagate the error down into chord and chain.
It's not perfect, and I only really tested it with this current example, but it's enough for me and a step in the right direction.

This is the code I needed to add: ```
class CeleryException(Exception):
    task_id = None
    task_name = None

    def __init__(self, task_id, task_name, exc):
        self.task_id = task_id
        self.task_name = task_name
        self.exc = exc

    def __str__(self):
        return 'Exception raised in celery taskId: ' + \
            self.task_id + '. TaskName: ' + self.task_name + '.' \
            'Exception: ' + str(self.exc)


class MfTask(Task):
    def __call__(self, *args, **kwargs):
        # If an error has been raised from a previous task, pass down the error
        for arg in args:
            # Need to handle list of chord
            if type(arg) is list:
                for el in arg:
                    if type(el) is CeleryException:
                        return el

            if type(arg) is CeleryException:
                return arg

        try:
            return Task.__call__(self, *args, **kwargs)
        except Exception as e:
            return CeleryException(self.request.id, self.name, e)


class MfAsyncResult(AsyncResult):
    # This needs to be done inside a class implementing AsyncResult
    def get(self, *args):
        res = AsyncResult.get(self, args)

        if type(res) is CeleryException:
            raise res

        return res
```

The test I ran:
```
@app.task(base=MfTask)
def add(x, y, failes):
    if failes:
        raise Exception('Failing in add')
    return x + y


@app.task(base=MfTask)
def mul(x, y, failes):
    if failes:
        raise Exception('Failing in mul')
    return x * y


@app.task(base=MfTask)
def sumall(v, failes):
    if failes:
        raise Exception('Failing in sumall')
    return sum(v)


@app.task(base=MfTask)
def some_test(main_f, add_f, sumall_f, mul_f):
    nbs = [1, 2, 3, 4]

    add_chord = chord(
        (add.s(nb, 1, add_f) for nb in nbs),
        sumall.s(sumall_f)
    )

    add_async = chain(
        add_chord, mul.s(2, mul_f)).apply_async()

    if main_f:
        raise Exception('Failing in some_test')

    return add_async.task_id


def test(main=False, add=False, sumall=False, mul=False):
    taskid = some_test.delay(
        main,
        add,
        sumall,
        mul
    ).get()

    res = MfAsyncResult(taskid).get()

    print('Res is', res, 'type is: ', type(res))
```

Running: ```
test(add=True)
```
Will yield: ```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/src/apps/celery_app/tasks.py", line 132, in test
    res = MfAsyncResult(taskid).get()
  File "/src/apps/celery_app/tasks.py", line 64, in get
    raise res
apps.celery_app.tasks.CeleryException: Exception raised in celery taskId: 9a8f0630-7f32-4975-8852-62e6831e9611. TaskName: apps.celery_app.tasks.add. Error: Failing in add
```

**Comment by auvipy** [comment]
is it still reproducable in master? will revisit after 4.2 release

**Comment by Lobstrio** [comment]
Hello, this is still the case. Reproduced exactly the same issue today.

**Comment by auvipy** [comment]
with celery 4.3 and celery master?

**Comment by Lobstrio** [comment]
> with celery 4.3 and celery master?

Yes confirmed. Just tested again with celery 4.3 and master branch

**Comment by SharpEdgeMarshall** [comment]
I don't know why but if you put a sleep(5) before the `res.get()` it works.

**Comment by Lobstrio** [comment]
@auvipy why closing this?

**Comment by Lobstrio** [comment]
@SharpEdgeMarshall is you first task in the chain executing in less than 5 seconds?

**Comment by auvipy** [comment]
I don't know why but if you put a sleep(5) before the res.get() it works.

**Comment by SharpEdgeMarshall** [comment]
> I don't know why but if you put a sleep(5) before the res.get() it works.

Are you joking? I was giving a workaround/hint about what could be the bug and you close the issue using my comment as a solution?

This is a critical issue, 2 days ago we had a lot of exception inside chains and our Redis was near to break down because of continuous `get()` summing up.

**Comment by Lobstrio** [comment]
@SharpEdgeMarshall  @auvipy  this is a real issue and should be reopened.
I didn't find convenient workaround so I just handle results within tasks without using celery backend.

**Comment by SharpEdgeMarshall** [comment]
I think I've found the solution (and is really stupid) I'm working on a PR

**Comment by auvipy** [comment]
Please try from master

**Comment by AnthonyDeroche** [comment]
I'm still encountering this issue in v4.4.0. When a task fails in a chain being part of a group, itself being the header of a chord, then ChordError is not raised, and the task hangs forever.

**Comment by auvipy** [comment]
https://github.com/celery/celery/pull/5587 didn't work then?

**Comment by AnthonyDeroche** [comment]
It works if you run only a simple chain. 
If you run two chains in a chord header, if a chain fails then the chord callback is not called and ChordError is not raised. The task which fails within the chain must not be the last to trigger the bug. 

I cannot work on a minimal code to reproduce the bug before january.

**Comment by auvipy** [comment]
I have added this to next minor release. please come with sample code in January

**Comment by wilcoschoneveld** [comment]
This is a minimal test case for this issue:

https://github.com/wilcoschoneveld/celery-docker-quickstart/tree/chain_error

Run worker with:
```
> docker-compose up
```

And in a different terminal:
```
> docker-compose exec worker python script.py
```

This will make a simple chain with two add tasks and wait for the result. The add task will throw a `ZeroDivisionError` and this is not propagated.

@auvipy

**Comment by auvipy** [comment]
> This is a minimal test case for this issue:
> 
> https://github.com/wilcoschoneveld/celery-docker-quickstart/tree/chain_error
> 
> Make sure to checkout the `chain_error` branch.
> 
> Run worker with:
> 
> ```
> > docker-compose up
> ```
> 
> And in a different terminal:
> 
> ```
> > docker-compose exec worker python script.py
> ```
> 
> This will make a simple chain with two add tasks and wait for the result. The add task will throw a `ZeroDivisionError` and this is not propagated.
> 
> @auvipy

did you find any fix for this? I was on vacation. sorry for the late response

**Comment by wilcoschoneveld** [comment]
@auvipy no, I did not.

**Comment by adrienyhuel** [comment]
I found that when you write a chord with a group of chains in the header, the tasks (in the chains) requests contains two arguments : "group" and "chord".

- For all tasks, chord contains thing like that : 
```
{'task': 'tasks.b1', 'args': [], 'kwargs': {}, 'options': {'task_id': '6a2d85a8-5c5c-47ff-b7fb-b3758098fe98', 'reply_to': '8f844223-89cc-363b-b80a-70911eeab115'}, 'subtask_type': None, 'chord_size': 2, 'immutable': False}, 'called_directly': False, '_protected': 1}
```

- For the last tasks of the chains, group contains the group_id, otherwise it is None.

When a task finish (failure or success), it checks if the arg "chord" is not None and call "on_chord_part_return" method.
So on all tasks, this method is called.
But in this method, it checks if "group" is not None, and then finish the chord (ie : call the callback if all chord header parts are finished).
So, the chord can only finish when all last tasks of all chains are finished.

When there is an error in the middle of a chain, the last task of the chain is never called, and the chord never finish.

I hope this can help :), I don't know how to go further.

**Comment by adrienyhuel** [comment]
@auvipy it seems that the problem is solved now.

I tried the sample from OP @zaro and using lastest Celery (5.1.1), RabbitMQ as broker and Redis as result backend.

No hanging anymore :)

With 5.0.5 it hangs, and not with 5.1.0. So it seems something was solved in 5.1.0
